File: codegen/src/main/scala/bsp/codegen/docs/MarkdownRenderer.scala
package bsp.codegen.docs

import bsp.codegen._
import bsp.codegen.ir.Hint.{Documentation, Deprecated}
import bsp.codegen.ir.JsonRPCMethodType.{Notification, Request}
import bsp.codegen.ir.{Def, Hint, Operation, PolymorphicDataKind}
import cats.syntax.all._
import software.amazon.smithy.model.shapes.ShapeId

import scala.collection.mutable.{Set => MSet}

object MarkdownRenderer {
  def render(tree: DocTree): String = {
    val visited = MSet.empty[ShapeId]
    val version = ProtocolVersionLoader.version()
    val renderer = new MarkdownRenderer(tree, visited, version)
    val rendered = renderer.render.render

    val notRenderedIds = visited.diff(tree.structures.keys.toSet)
    notRenderedIds.foreach { id =>
      println(s"warning: ${id.getName} was not rendered")
    }

    rendered
  }
}

class MarkdownRenderer private (tree: DocTree, visited: MSet[ShapeId], protocolVersion: String) {
  import bsp.codegen.Settings.typescript
  import dsl._

  val tsRenderer = new TypescriptRenderer

  def render: Lines = {
    val commonShapes =
      if (tree.commonShapeIds.isEmpty) empty
      else
        lines(
          "## Common shapes",
          newline,
          tree.commonShapeIds.foldMap(renderStructureNode),
          newline
        )

    // Server comes first because it's bigger.
    val services = tree.services.sortBy(_.operations.size).reverse.foldMap(renderServiceNode)

    val dataKinds = tree.dataKindInhabitants.toList.foldMap { case (extendableType, kinds) =>
      renderExtendableTypeNode(extendableType, kinds)
    }

    val versionRendered = lines(
      "## BSP version",
      s"`$protocolVersion`",
      newline
    )

    lines(
      versionRendered,
      commonShapes,
      services,
      dataKinds
    )
  }

  def renderStructureNode(id: ShapeId): Lines = {
    // We don't want to generate definitions for built-in primitives.
    if (id.getNamespace == "smithy.api") {
      return empty
    }

    val node = tree.structures(id)

    if (visited.contains(id)) empty
    else {
      visited.add(id)
      lines(
        renderStructureNodeDef(node.definition),
        node.definition.members
          .filter(_.getNamespace == id.getNamespace)
          .foldMap(renderStructureNode)
      )
    }
  }

  def renderStructureNodeDef(definition: Def): Lines = {
    lines(
      s"#### ${definition.shapeId.getName()}",
      newline,
      documentation(definition.hints),
      newline,
      tsBlock(definition),
      newline
    )
  }

  def renderExtendableTypeNode(extendableType: ShapeId, kinds: List[PolymorphicDataKind]): Lines = {
    val header = s"## ${extendableType.getName()} kinds"
    val kindNodes = kinds.foldMap(renderDataKindNode(extendableType))
    lines(
      header,
      newline,
      kindNodes
    )
  }

  def renderDataKindNode(extendableType: ShapeId)(kind: PolymorphicDataKind): Lines = {
    lines(
      s"### ${kind.shapeId.getName()}",
      "This structure is embedded in",
      s"the `data?: ${extendableType.getName()}` field, when",
      s"""the `dataKind` field contains `"${kind.kind}"`.""",
      newline,
      renderStructureNode(kind.shapeId)
    )
  }

  def renderOperationNode(node: OperationDocNode): Lines = {
    val OperationDocNode(operation, input, output) = node
    lines(
      s"### ${operation.name}: ${methodTpe(operation)}",
      newline,
      documentation(operation.hints),
      newline,
      s"- method: `${operation.jsonRPCMethod}`",
      input.foldMap(n => s"- params: `${n.getName()}`"),
      output.foldMap(n => s"- result: `${n.getName()}`"),
      newline,
      input.foldMap(renderStructureNode),
      output.foldMap(renderStructureNode)
    )
  }

  def renderServiceNode(node: ServiceDocNode): Lines = {
    val ServiceDocNode(shapeId, operations) = node
    val header = if (shapeId.getName().toLowerCase().contains("server")) {
      s"## BSP Server remote interface"
    } else {
      s"## BSP Client remote interface"
    }
    lines(
      header,
      newline,
      operations
        .map(tree.operations)
        .foldMap(renderOperationNode)
    )
  }

  def methodTpe(operation: Operation): String = operation.jsonRPCMethodType match {
    case Notification => "notification"
    case Request      => "request"
  }

  def documentation(hints: List[Hint]): Lines = Lines {
    val maybeDeprecated = hints.collectFirst { case Deprecated(message) =>
      val line = if (message.isEmpty) List("**Deprecated**") else List(s"**Deprecated**: $message")
      lines(
        line,
        newline
      )
    }

    val docs = hints.collect { case Documentation(string) =>
      string.split(System.lineSeparator()).toList
    }

    lines(
      maybeDeprecated,
      docs
    )
  }

  def tsBlock(definition: Def): Lines =
    tsRenderer
      .render(definition)
      .map { tsCode =>
        lines(
          "```ts",
          tsCode,
          "```"
        )
      }
      .getOrElse(empty)

}

File: codegen/src/main/kotlin/bsp/codegen/docs/MarkdownRenderer.kt
package bsp.codegen.docs

import bsp.codegen.common.Loader
import bsp.codegen.common.dsl.CodeBlock
import bsp.codegen.common.dsl.code
import bsp.codegen.common.dsl.emptyCode
import bsp.codegen.common.ir.Def
import bsp.codegen.common.ir.Hint
import bsp.codegen.common.ir.JsonRpcMethodType
import bsp.codegen.common.ir.Operation
import bsp.codegen.common.ir.PolymorphicDataKind
import bsp.codegen.common.ir.Type
import bsp.codegen.common.ir.referencedShapeIds
import software.amazon.smithy.model.shapes.ShapeId

class MarkdownRenderer(
    val tree: DocTree,
    val visited: MutableSet<ShapeId>,
    val protocolVersion: String
) {
  companion object {
    fun render(tree: DocTree): String {
      val visited = mutableSetOf<ShapeId>()
      val version = Loader.protocolVersion
      val renderer = MarkdownRenderer(tree, visited, version)
      val rendered = renderer.render().toString()

      val notRenderedIds = visited.minus(tree.structures.keys.toSet())
      notRenderedIds.forEach { println("warning: ${it.name} was not rendered") }

      return rendered
    }
  }

  val tsRenderer = TypescriptRenderer()

  fun render(): CodeBlock {
    val commonShapes =
        if (tree.commonShapeIds.isEmpty()) emptyCode
        else
            code {
              -"## Common shapes"
              newline()
              tree.commonShapeIds.forEach { include(renderStructureNode(it)) }
              newline()
            }

    // Server comes first because it's bigger.
    val services = code {
      tree.services
          .sortedBy { it.operations.size }
          .reversed()
          .forEach { include(renderServiceNode(it)) }
    }

    val dataKinds = code {
      tree.dataKindInhabitants.forEach { (extendableType, kinds) ->
        include(renderExtendableTypeNode(extendableType, kinds))
      }
    }

    val versionRendered = code {
      -"## BSP version"
      -"`$protocolVersion`"
      newline()
    }

    return code {
      include(versionRendered)
      include(commonShapes)
      include(services)
      include(dataKinds)
    }
  }

  fun renderStructureNode(id: ShapeId): CodeBlock {
    // We don't want to generate definitions for built-in primitives.
    if (id.namespace == "smithy.api") {
      return emptyCode
    }

    val node =
        tree.structures[id]
            ?: throw RuntimeException("Id not found: $id, structures: ${tree.structures}")

    if (visited.contains(id)) return emptyCode

    visited.add(id)
    return code {
      include(renderStructureNodeDef(node.definition))
      node.definition
          .referencedShapeIds()
          .filter { it.namespace == id.namespace }
          .forEach { include(renderStructureNode(it)) }
    }
  }

  fun renderStructureNodeDef(definition: Def): CodeBlock = code {
    -"#### ${definition.shapeId.name}"
    newline()
    include(documentation(definition.hints))
    newline()
    include(tsBlock(definition))
    newline()
  }

  fun renderExtendableTypeNode(
      extendableType: ShapeId,
      kinds: List<PolymorphicDataKind>
  ): CodeBlock = code {
    -"## ${extendableType.name} kinds"
    newline()
    kinds.forEach { include(renderDataKindNode(extendableType)(it)) }
  }

  fun renderDataKindNode(extendableType: ShapeId): (kind: PolymorphicDataKind) -> CodeBlock =
      { kind ->
        code {
          -"### ${kind.shape.name}"
          -"This structure is embedded in"
          -"the `data?: ${extendableType.name}` field, when"
          -"the `dataKind` field contains `\"${kind.kind}\"`."
          newline()
          include(renderStructureNode(kind.shape))
        }
      }

  fun renderOperationNode(node: OperationDocNode): CodeBlock = code {
    -"### ${node.operation.name}: ${methodTpe(node.operation)}"
    newline()
    include(documentation(node.operation.hints))
    newline()
    -"- method: `${node.operation.jsonRpcMethod}`"
    node.operation.inputType
        .takeIf { it != Type.Unit }
        ?.let { -"- params: `${tsRenderer.renderType(it)}`" }
    node.operation.outputType
        .takeIf { it != Type.Unit }
        ?.let { -"- result: `${tsRenderer.renderType(it)}`" }
    newline()
    (node.operation.inputType as? Type.Ref)?.let { include(renderStructureNode(it.shapeId)) }
    (node.operation.outputType as? Type.Ref)?.let { include(renderStructureNode(it.shapeId)) }
  }

  fun renderServiceNode(node: ServiceDocNode): CodeBlock = code {
    if (node.shapeId.name.lowercase().contains("server")) {
      -"## BSP Server remote interface"
    } else {
      -"## BSP Client remote interface"
    }
    newline()
    node.operations.forEach { include(renderOperationNode(tree.operations[it]!!)) }
  }

  fun methodTpe(operation: Operation): String =
      when (operation.jsonRpcMethodType) {
        JsonRpcMethodType.Request -> "request"
        JsonRpcMethodType.Notification -> "notification"
      }

  fun documentation(hints: List<Hint>): CodeBlock = code {
    val maybeDeprecated =
        hints
            .firstNotNullOfOrNull { it as? Hint.Deprecated }
            ?.let { deprecated ->
              if (deprecated.message.isNotEmpty()) {
                "**Deprecated**: ${deprecated.message}"
              } else {
                "**Deprecated**"
              }
            }

    val maybeUnstable =
        "**Unstable** (may change in future versions)\n"
            .takeIf { hints.any { it is Hint.Unstable } }

    val docs =
        hints.filterIsInstance<Hint.Documentation>().flatMap {
          it.string.split(System.lineSeparator())
        }

    -maybeDeprecated
    -maybeUnstable
    newline()
    docs.forEach { -it }
  }

  fun tsBlock(definition: Def): CodeBlock = code {
    tsRenderer.render(definition)?.let { tsCode ->
      -"```ts"
      include(tsCode)
      -"```"
    }
  }
}
